addons/mail/models/mail_scheduled_message.py

```diff
@@ -0,0 +1,262 @@
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _("Edit Scheduled Note") if self.is_note else _("Edit Scheduled Message"),
+            'res_model': self._name,
+            'view_mode': 'form',
+            'views': [[False, 'form']],
+            'target': 'new',
+            'res_id': self.id,
+        }
+
+    def post_message(self):
+        self.ensure_one()
+        if self.env.is_admin() or self.create_uid.id == self.env.uid:
+            self._post_message()
+        else:
+            raise UserError(_("You are not allowed to send this scheduled message"))
+
+    def _post_message(self, raise_exception=True):
+        """ Post the scheduled messages.
+            They are posted using their creator as user so that one can check that the creator has
+            still post permission on the related record, and to allow for the attachments to be
+            transferred to the messages (see _process_attachments_for_post in mail.thread)
+            if raise_exception is set to False, the method will skip the posting of a message
+            instead of raising an error. This is useful when scheduled messages are sent from
+            the _post_messages_cron
+        """
+        notification_parameters_whitelist = self._notification_parameters_whitelist()
+        for scheduled_message in self:
+            message_creator = scheduled_message.create_uid
+            try:
+                scheduled_message.with_user(message_creator)._check()
+            except AccessError:
+                if raise_exception:
+                    raise
+                _logger.info("Posting of scheduled message %s failed: user %s cannot post on the record", scheduled_message.id, message_creator.id)
+                continue
+            self.env[scheduled_message.model].browse(scheduled_message.res_id).with_user(message_creator).message_post(
+                attachment_ids=list(scheduled_message.attachment_ids.ids),
+                author_id=scheduled_message.author_id.id,
+                body=scheduled_message.body,
+                partner_ids=list(scheduled_message.partner_ids.ids),
+                subtype_xmlid='mail.mt_note' if scheduled_message.is_note else 'mail.mt_comment',
+                **{k: v for k, v in json.loads(scheduled_message.notification_parameters or '{}').items() if k in notification_parameters_whitelist},
+            )
+        self.unlink()
+
+    # ------------------------------------------------------
+    # Business Methods
+    # ------------------------------------------------------
+
+    @api.model
+    def _check(self, values=None):
+        """ Restrict the access to a scheduled message.
+            Access is based on the record on which the scheduled message will be posted to
+        """
+        if self.env.is_superuser():
+            return True
+
+        model_ids = defaultdict(set)
+        if self:
+            for scheduled_message in self:
```

Comment:
If we want to avoid a serpent qui se mord la queue, shouldn't we access data as sudo ?
```suggestion
            for scheduled_message in self.sudo():
```